### PR TITLE
ext-foreign-toplevel-list: Implement protocol

### PR DESCRIPTION
**Pull request by columbarius**
Requires: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4543

### CODE REVIEW

File: sway/tree/view.c [hunk 1]
```diff
@@ -828,6 +834,10 @@ void view_map(struct sway_view *view, struct wlr_surface *wlr_surface,
 		input_manager_set_focus(&view->container->node);
 	}
 
+	foreign_toplevel_state.app_id = (char *)view_get_app_id(view);
+	foreign_toplevel_state.title = (char *)view_get_class(view);
```
Comment (columbarius) [inline sway/tree/view.c:838]: Is this ok, or should the wlroots struct use the const qualifier?

Comment (emersion) [reply sway/tree/view.c:838]: IMHO would be better if the wlroots struct used the const qualifier

File: sway/tree/view.c [hunk 2]
```diff
@@ -751,6 +751,12 @@ void view_map(struct sway_view *view, struct wlr_surface *wlr_surface,
 		}
 	}
 
+	struct wlr_ext_foreign_toplevel_handle_v1_state foreign_toplevel_state = {0};
+	view->ext_foreign_toplevel =
+		wlr_ext_foreign_toplevel_handle_v1_create(server.foreign_toplevel_list, &foreign_toplevel_state);
```
Comment (emersion) [inline sway/tree/view.c:756]: Hm, it's not nice to create the foreign toplevel handle with an empty initial state.

Is it too early to call `view_get_app_id()` here?

File: sway/tree/view.c [hunk 3]
```diff
@@ -1060,6 +1075,14 @@ void view_update_title(struct sway_view *view, bool force) {
 	if (view->foreign_toplevel && title) {
 		wlr_foreign_toplevel_handle_v1_set_title(view->foreign_toplevel, title);
 	}
+
+	if (view->ext_foreign_toplevel) {
+		struct wlr_ext_foreign_toplevel_handle_v1_state foreign_toplevel_state = {
+			.app_id = (char *)view_get_app_id(view),
+			.title = (char *)title,
+		};
+		wlr_ext_foreign_toplevel_handle_v1_update_state(view->ext_foreign_toplevel, &foreign_toplevel_state);
```
Comment (emersion) [inline sway/tree/view.c:1084]: Suggestion: maybe we can extract this into a helper function?

**Comment by emersion** [comment]
We also need to update the state when the app_id changes.

**Comment by columbarius** [comment]
> We also need to update the state when the app_id changes.

This is currently not done for the toplevel_management protocol, right?

**Comment by emersion** [comment]
>      This is currently not done for the toplevel_management protocol, right?

Correct, the old impl only sets app_id when mapping and then ignores any changes.

File: sway/tree/view.c (hunk 1 above)
Comment (columbarius) [reply sway/tree/view.c]: Done

File: sway/tree/view.c (hunk 2 above)
Comment (columbarius) [reply sway/tree/view.c:767]: Testing showed we can create the handle with app_id and title!

File: sway/tree/view.c (hunk 3 above)
Comment (columbarius) [reply sway/tree/view.c]: Done

**Comment by columbarius** [comment]
I've added a function updating the app_id analog to the title.

File: sway/tree/view.c [hunk 4]
```diff
@@ -582,6 +582,14 @@ static struct sway_workspace *select_workspace(struct sway_view *view) {
 	return NULL;
 }
 
+static void update_ext_foreign_toplevel(struct sway_view *view) {
+	struct wlr_ext_foreign_toplevel_handle_v1_state toplevel_state;
+	toplevel_state.app_id = view_get_app_id(view);
+	toplevel_state.title = view_get_title(view);
+	sway_log(SWAY_ERROR, "Update ext foreign topleven: app_id %s, title %s", toplevel_state.app_id, toplevel_state.title);
```
Comment (emersion) [inline sway/tree/view.c:589]: Remnant from a debugging session?

File: sway/tree/view.c [hunk 5]
```diff
@@ -582,6 +582,14 @@ static struct sway_workspace *select_workspace(struct sway_view *view) {
 	return NULL;
 }
 
+static void update_ext_foreign_toplevel(struct sway_view *view) {
+	struct wlr_ext_foreign_toplevel_handle_v1_state toplevel_state;
+	toplevel_state.app_id = view_get_app_id(view);
+	toplevel_state.title = view_get_title(view);
```
Comment (emersion) [inline sway/tree/view.c:588]: Please use initializers when possible: this ensures there are no undefined memory if a new field is added to the struct in a new wlroots version.

```c
const struct wlr_ext_foreign_toplevel_handle_v1_state toplevel_state = {
    .app_id = view_get_app_id(view),
    .title = view_get_title(view),
};
```

**Comment by emersion** [comment]
With these fixed, LGTM!

File: sway/tree/view.c (hunk 4 above)
Comment (columbarius) [reply sway/tree/view.c]: Ok, removed to logging.

File: sway/tree/view.c (hunk 5 above)
Comment (columbarius) [reply sway/tree/view.c]: Done

**Review by emersion — APPROVED** [review]
Thanks!

**Comment by columbarius** [comment]
Thanks for merging!

**Pull request merged**